[4348] Sidebar event propagation between widgets seems not to work

aachavarria opened the pull request:
https://github.com/craftercms/craftercms/issues/4348

aachavarria committed:
adding copy And paste Notifications; Supporting Duplicate;

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4348

aachavarria committed:
adding copy, duplicate, contentTypeChanged notification to legacy;

aachavarria committed:
Renaming Change Template to Change Type;
ContentTypesFilter.tsx: showing selected filter;
NewContentDialog.tsx: adding default filter when the type of the dialog is 'change', using defaultFilter to filter the initial List;

aachavarria committed:
adding keyword reset on catalogReset

aachavarria committed:
adding itemPasted event;
pathNavigator.ts: listening itemPasted event;

aachavarria committed:
Adding ITEM_DELETED event;

rart commented on static-assets/components/cstudio-contextual-nav/wcm-site-dropdown-mods/wcm-root-folder.js:
```diff
@@ -977,7 +977,7 @@
           markup = markup.replace(/\&lt;/g, '<');
           markup = markup.replace(/\&amp;/g, '&');
-          markup = markup.replace(/\&#x27;/g, "'");
+          markup = markup.replace(/\&#x27;/g, '\'');
```
The `&#x27;` is the `'` character. Why this change?

aachavarria replied:
Prettier

rart commented on static-assets/components/cstudio-contextual-nav/wcm-site-dropdown-mods/wcm-root-folder.js:
```diff
@@ -1512,7 +1512,8 @@
                   },
-                  failure: function() {}
+                  failure: function() {
+                  }
```
All these changes look like IntelliJ formatter

aachavarria replied:
looks like there was a issue with prettier, will run it again

rart commented on ui/app/src/components/ItemMenu/ItemMenu.tsx:
```diff
@@ -329,7 +331,13 @@ export function ItemMenu(props: ItemMenuProps) {
+              type: 'ITEM_PASTED',
+              payload: {
+                path: item.path
+              }
```
let's go ahead and send the whole item

rart commented on ui/app/src/modules/Content/Authoring/NewContentDialog.tsx:
```diff
@@ -209,7 +209,7 @@ function NewContentDialogBody(props: NewContentDialogProps) {
   } = props;
   const [openSelector, setOpenSelector] = useState(false);
-  const defaultFilterType = 'all';
+  const defaultFilterType = (type === 'new') ? 'all' : item.systemType;
```
I think we **may be** masking a backend bug. Perhaps we should roll this back and display strictly as the backend replies.

aachavarria replied:
I don't think so, because the BE service is for creating content in the selected path, not for changing the selected path contentType

rart commented on ui/app/src/modules/Content/Delete/DeleteDialog.tsx:
```diff
@@ -309,6 +311,14 @@ function DeleteDialogWrapper(props: DeleteDialogProps) {
+        hostToHost$.next({
+          type: 'ITEM_DELETED',
+          payload: {
+            paths: selectedItems
```
Can we send also like the main item?

rart commented on ui/app/src/services/content.ts:
```diff
@@ -951,7 +951,7 @@ export function getChildrenByPath(
         parent: parseLegacyItemToSandBoxItem(item),
-        ...(item.children[0].contentType === '/component/level-descriptor' && {
+        ...(item.children[0]?.contentType === '/component/level-descriptor' && {
           levelDescriptor: {
```
- This prop should always be there. If the level descriptor isn't there, leave it in as null.
- Perhaps we should be finding the item instead of assuming is the first.

rart commented on ui/app/src/utils/i18n-legacy.ts:
```diff
@@ -1396,9 +1396,17 @@ export const itemSuccessMessages = defineMessages({
+  },
+  itemContentTypeChanged: {
+    id: 'item.contentTypeChanged',
+    defaultMessage: 'Item content type changed successfully'
```
item ~content~ type changed successfully

aachavarria committed:
Running Prettier

aachavarria committed:
Updates

aachavarria committed:
adding new events propagation for FOLDER_RENAMED and FOLDER_CREATED

aachavarria committed:
Adding new Dialog changeContentType,
Changes on NewContentDialog.tsx to improve logic;
adding new event propagation for ITEM_CREATED

aachavarria committed:
Updating SearchBar showActionButton;
adding new Dialog ChangeContentType;

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4348

aachavarria committed:
adding skeleton for NewContentCard;

aachavarria committed:
removing timeout

aachavarria committed:
embedded_legacy_form.ftl: adding new prop, isNew to check if the for is for a newFile;
BulkUploadDialog.tsx: adding FILE_UPLOADED event propagation;
CreateFileDialog.tsx: adding FILE_CREATED event propagation;
CreateFolderDialog.tsx.tsx: adding FILE_CREATED event propagation;
LegacyFormDialog.tsx: adding ITEM_CREATED and ITEM_UPDATED event propagation;
adding systemEvents, listening systemEvents.ts on PathNavigator.tsx;

rart commented on ui/app/src/components/Dialogs/BulkUploadDialog.tsx:
```diff
@@ -641,6 +651,10 @@ export default function BulkUploadDialog(props: BulkUploadProps) {
 
+  const onFileUploaded = (path: string) => {
+    dispatch(systemEvent(fileUploaded({ target: path })));
+  };
```
Will trigger effect on every render

rart commented on ui/app/src/components/ItemMenu/ItemMenu.tsx:
```diff
@@ -240,15 +236,13 @@ export function ItemMenu(props: ItemMenuProps) {
-                  closeNewContentDialog(),
-                  changeContentType({ contentTypeId: item.contentTypeId, path })
+                  closeChangeContentTypeDialog(),
+                  changeContentType({ contentTypeId: item.contentTypeId, path: item.path })
```
Can we have `newContentTypeId` and `originalContentTypeId`?

aachavarria replied:
yes

rart commented on ui/app/src/services/content.ts:
```diff
@@ -1002,7 +1002,7 @@ export function duplicate(site: string, path: string): Observable<string> {
     switchMap(() => paste(site, getParentPath(parentPath))),
     // Duplicate only does shallow copy, so return the copied path
-    pluck('status', '0')
+    map((paths) => paths[0])
```
`pluck('0')`

rart commented on ui/app/src/state/actions/dialogs.ts:
```diff
@@ -200,10 +210,23 @@ export const pathSelectionDialogClosed = createAction('PATH_SELECTION_CLOSED');
+);
+export const showRevertItemSuccessNotification = createAction(
+  'SHOW_REVERT_ITEM_SUCCESS_NOTIFICATION'
+);
```
Let's group all these system stuff under the new systemEvents actions file, only that, rename actions/systemEvents.ts to actions/system.ts

rart commented on ui/app/src/state/actions/systemEvents.ts:
```diff
@@ -0,0 +1,40 @@
+/*
```
Rename to system.ts

rart commented on ui/app/src/state/actions/systemEvents.ts:
```diff
@@ -0,0 +1,40 @@
+  'ITEM_DUPLICATED'
+);
+
+export const systemEvent = createAction<StandardAction>('SYSTEM_EVENT');
```
let's call this one emitSystemEvent

rart commented on ui/app/src/state/epics/dialogs.ts:
```diff
@@ -117,76 +120,139 @@ export default [
+    ),
+  (action$, state$, { intlRef: { current: intl } }: { intlRef: { current: IntlShape } }) =>
+    action$.pipe(
+      ofType(showCutItemSuccessNotification.type),
```
Let's add all these to epics/system.ts

rart commented on ui/app/src/state/epics/misc.ts:
```diff
@@ -15,65 +15,77 @@
-    return NEVER;
-  })
-);
+const systemEventPropagate: Epic = (action$) =>
```
==> epics/system.ts

rart reviewed (changes requested):
Please address comments and fix conflicts

aachavarria committed:
PR Review Changes;
removing Delete, Cut and ChangeContentType options from ItemMenu on RootFolders;

aachavarria committed:
moving emitSystemEvent to system.ts;

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4348

 Conflicts:
	ui/app/src/modules/Content/Authoring/NewContentDialog.tsx
	ui/app/src/modules/System/Sites/Create/CreateSiteDialog.tsx

aachavarria committed:
adding skeleton to PathNavigator and ItemMenu;

aachavarria committed:
Typo

aachavarria committed:
CopyDialog style fixes;
changes on clipboard state;

aachavarria committed:
template-editor: changing position of cancel btn;
all Dialogs: fixing style of dark theme;

aachavarria committed:
PathNavigator Updates to use new action pathNavigatorConditionallySetPath;

aachavarria committed:
missing prop

rart commented on static-assets/components/cstudio-forms/template-editor.js:
```diff
@@ -65,7 +65,7 @@ CStudioAuthoring.Module.requireModule(
                 'transform-path-to-url': {
                   label: 'Transform PATH to URL',
-                  value: '${urlTransformationService.transform(\'storeUrlToRenderUrl\', STOREURL)}'
+                  value: "${urlTransformationService.transform('storeUrlToRenderUrl', STOREURL)}"
```
revert.
something seemed to have happened with this file. seeing to many quote changes that go against the format setting. any idea what’s up?

rart replied:
ha! Found it: https://prettier.io/docs/en/next/rationale.html#strings
Hadn't seen this until now, perhaps is recent.

rart commented on ui/app/src/components/Dialogs/BulkUploadDialog.tsx:
```diff
@@ -641,6 +649,13 @@ export default function BulkUploadDialog(props: BulkUploadProps) {
 
+  const onFileUploaded = useCallback(
+    (path: string) => {
+      dispatch(emitSystemEvent(fileUploaded({ target: path })));
```
Do we also have a fileCreated? Is there a difference between upload and creation? Perhaps we should only leave the fileCreated

aachavarria replied:
will change to use fileCreated

rart commented on ui/app/src/components/Navigation/PathNavigator/PathNavigator.tsx:
```diff
@@ -31,35 +31,51 @@ import {
 import Breadcrumbs from './PathNavigatorBreadcrumbs';
 import NavItem from './PathNavigatorItem';
 import Nav from './PathNavigatorList';
-import ContentLoader from 'react-content-loader';
```
yarn remove it too

rart commented on ui/app/src/components/Navigation/PathNavigator/PathNavigator.tsx:
```diff
@@ -31,35 +31,51 @@ import {
+  itemUpdated
+} from '../../../state/actions/system';
+import List from '@material-ui/core/List';
+import { PathNavigatorSkeletonItem } from './PathNavigatorSkeletonItem';
```
Should be default export

rart commented on ui/app/src/models/GlobalState.ts:
```diff
@@ -98,7 +99,11 @@ export interface GlobalState {
+    clipboard: {
+      type: 'cut | copy';
+      paths: string[];
+      sourceRootPath: string;
```
This is the path where things were copied/cutter from, right? If so, let’s call it `sourcePath`.

Also, please move the workflow check from cut to paste – when type is cut.

aachavarria replied:
it is the rootPath it is used to check if the itemMenu should show the paste option;

rart commented on ui/app/src/modules/Content/History/HistoryDialog.tsx:
```diff
@@ -53,6 +52,7 @@ import Dialog from '@material-ui/core/Dialog';
 import { batchActions } from '../../../state/actions/misc';
 import { fetchUserPermissions } from '../../../state/actions/content';
 import { asDayMonthDateTime } from '../../../utils/datetime';
+import { makeStyles } from '@material-ui/core';
```
fix core import

rart commented on ui/app/src/modules/Content/History/VersionList.tsx:
```diff
@@ -30,11 +29,12 @@ import { LegacyVersion } from '../../../models/Version';
 import clsx from 'clsx';
 import palette from '../../../styles/palette';
 import LookupTable from '../../../models/LookupTable';
+import { makeStyles } from '@material-ui/core';
```
core import

rart commented on ui/app/src/state/actions/dialogs.ts:
```diff
@@ -101,6 +103,16 @@ export const closeNewContentDialog = createAction<StandardAction>('CLOSE_NEW_CON
+export const closeChangeContentTypeDialog = createAction<StandardAction>(
+  'CLOSE_CHANGE_CONTENT_TYPE_DIALOG'
+);
+export const ChangeContentTypeDialogClosed = createAction('CHANGE_CONTENT_TYPE_DIALOG_CLOSED');
```
name out of pattern/convention

rart commented on ui/app/src/state/reducers/content.ts:
```diff
@@ -83,7 +101,9 @@ const reducer = createReducer<ContentState>(initialState, {
         ...state.items.permissionsByPath,
-        [payload.path]: createPresenceTable(payload.permissions.map(value => value.replaceAll(' ', '_').replace(/-/g, '_')))
+        [payload.path]: createPresenceTable(
+          payload.permissions.map((value) => value.replaceAll(' ', '_').replace(/-/g, '_'))
```
Consolidate into a single replace with regex

aachavarria committed:
Prettier format and PR review changes;

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4348

 Conflicts:
	static-assets/components/cstudio-common/common-api.js
	static-assets/components/cstudio-forms/template-editor.js
	ui/app/src/components/Dialogs/BulkUploadDialog.tsx
	ui/app/src/components/Dialogs/CopyDialog.tsx
	ui/app/src/components/Dialogs/CreateFolderDialog.tsx
	ui/app/src/components/ItemMenu/ItemMenu.tsx
	ui/app/src/components/Navigation/PathNavigator/PathNavigator.tsx
	ui/app/src/modules/Content/Authoring/NewContentDialog.tsx
	ui/app/src/modules/Content/Dependencies/DependenciesDialog.tsx
	ui/app/src/modules/Content/History/HistoryDialog.tsx
	ui/app/src/modules/Content/History/VersionList.tsx
	ui/app/src/modules/Content/Publish/PublishForm.tsx
	ui/app/src/modules/Search/Search.tsx
	ui/app/src/services/content.ts
	ui/app/src/state/actions/content.ts
	ui/app/src/state/actions/dialogs.ts
	ui/app/src/state/actions/misc.ts
	ui/app/src/state/actions/pathNavigator.ts
	ui/app/src/state/epics/dialogs.ts
	ui/app/src/state/epics/itemVersions.ts
	ui/app/src/state/epics/misc.ts
	ui/app/src/state/epics/pathNavigator.ts
	ui/app/src/state/epics/root.ts
	ui/app/src/state/reducers/content.ts
	ui/app/src/state/reducers/dialogs/createFile.ts
	ui/app/src/state/reducers/dialogs/createFolder.ts
	ui/app/src/utils/path.ts

aachavarria committed:
Prettier format and Merge Changes;

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4348

rart commented on ui/app/src/components/ItemMenu/ItemMenu.tsx:
```diff
@@ -243,20 +237,7 @@ export function ItemMenu(props: ItemMenuProps) {
-        cut(site, item).subscribe(
-          ({ success }) => {
-            if (success) {
-              dispatch(batchActions([setClipBoard({ path: item.path }), showCopyItemSuccessNotification()]));
```
Showing copy message instead of cut

rart commented on ui/app/src/components/ItemMenu/ItemMenu.tsx:
```diff
@@ -243,20 +237,7 @@ export function ItemMenu(props: ItemMenuProps) {
-            );
-          }
-        );
+        dispatch(itemCut({ path }));
```
Should be dispatched on success

rart commented on ui/app/src/components/ItemMenu/ItemMenu.tsx:
```diff
@@ -303,18 +296,22 @@ export function ItemMenu(props: ItemMenuProps) {
         break;
       }
       case 'paste': {
-        paste(site, item.path).subscribe(
```
Don't see the paste operation getting executed at all if this is removed. Looks like you're dispatching the paste system event before the paste is actually performed

aachavarria replied:
itemCut and itemPasted on this case arent actions from systemEvents;

rart commented on ui/app/src/components/ItemMenu/ItemMenu.tsx:
```diff
@@ -425,10 +422,7 @@ export function ItemMenu(props: ItemMenuProps) {
-              closeCreateFileDialog(),
-              onItemMenuActionSuccessCreator?.({ item, option: 'refresh' })
-            ])
+            onCreated: batchActions([closeCreateFileDialog()])
```
No longer batch

rart commented on ui/app/src/components/ItemMenu/ItemMenu.tsx:
```diff
@@ -438,10 +432,7 @@ export function ItemMenu(props: ItemMenuProps) {
-              closeCreateFileDialog(),
-              onItemMenuActionSuccessCreator?.({ item, option: 'refresh' })
-            ])
+            onCreated: batchActions([closeCreateFileDialog()])
```
No longer batch

rart commented on ui/app/src/components/ItemMenu/ItemMenu.tsx:
```diff
@@ -471,7 +462,7 @@ export function ItemMenu(props: ItemMenuProps) {
             path: item.path,
             site,
-            onClose: batchActions([closeUploadDialog(), onItemMenuActionSuccessCreator?.({ item, option: 'upload' })])
+            onClose: batchActions([closeUploadDialog()])
```
No longer batch

There's a bunch of these. Please fix all.

rart commented on ui/app/src/components/Navigation/PathNavigator/PathNavigator.tsx:
```diff
@@ -369,15 +427,13 @@ export default function PathNavigator(props: WidgetProps) {
 
 export function PathNavigatorUI(props: WidgetUIProps) {
   const classes = useStyles();
-  // region consts {...} = props
```
Leave region unless it ends up being a single line. If the region is removed, the endregion should be too.

rart commented on ui/app/src/components/Navigation/PathNavigator/PathNavigatorSkeletonItem.tsx:
```diff
@@ -0,0 +1,34 @@
+
+import { useStyles } from './styles';
+import ListItem from '@material-ui/core/ListItem';
+import { Skeleton } from '@material-ui/lab';
```
Check all lab imports to be loaded from its individual file (e.g. material/lab/skeleton)

rart commented on ui/app/src/modules/Content/Authoring/NewContentDialog.tsx:
```diff
@@ -158,58 +121,32 @@ export interface NewContentDialogStateProps extends NewContentDialogBaseProps {
   );
 }
 
-function NewContentDialogBody(props: NewContentDialogProps) {
```
Looks like the merge resulted in some changes getting lost. These renames should persist. ...DialogBody is the correct name.

aachavarria committed:
PR Review Changes;

aachavarria committed:
Optimize imports;

rart commented on ui/app/src/components/ToolsPanelPage/ToolsPanelPage.tsx:
```diff
@@ -18,6 +18,8 @@ import React, { useContext } from 'react';
 import ToolPanel from '../../modules/Preview/Tools/ToolPanel';
 import { renderWidgets, TempTestContext } from '../Widget/Widget';
 
+//console.log(formatMessage({ id: 'words.cancel', defaultMessage: 'words.cancel' }));
```
Should not commit

rart commented on ui/app/src/models/GlobalState.ts:
```diff
@@ -115,6 +122,7 @@ export interface GlobalState {
     currentUrl: string;
     computedUrl: string;
     showToolsPanel: boolean;
+    //toolsPanelPageStack>>pages[]
```
should not commit

rart commented on ui/app/src/models/GlobalState.ts:
```diff
@@ -178,6 +187,7 @@ export interface GlobalState {
     error: ApiResponse;
     isFetching: boolean;
     currentSite: string;
+    //preview>toolsPanel>widgets
```
should not commit

aachavarria committed:
removing comments;

aachavarria committed:
adding close action to Snackbars;

aachavarria committed:
Fixes issues related with the ItemMenu

rart commented on ui/app/src/components/CrafterCMSNextBridge.tsx:
```diff
@@ -104,3 +107,13 @@ function setUpLocaleChangeListener(update: Function, currentIntl: IntlShape) {
   return () => document.removeEventListener('setlocale', handler, false);
 }
+
+function SnackbarCloseButton({ id }) {
```
Move to it's own folder/file

aachavarria committed:
moving SnackbarCloseButton.tsx to it own file;

rart commented on ui/app/src/components/SnackbarCloseButton/SnackbarCloseButton.tsx:
```diff
@@ -0,0 +1,30 @@
+import CloseIcon from '@material-ui/icons/Close';
+import React from 'react';
+
+export default function SnackbarCloseButton({ id }) {
```
Props type please

rart replied:
Done.

rart committed:
Add props object typings

rart committed:
Merge branch 'develop' into feature/4348

rart reviewed (approved):
Conflicts resolved.
